File: engine/core/conjunction.py
from __future__ import annotations

import math
from dataclasses import dataclass, field
from enum import StrEnum
from typing import List
import numpy as np

from .propagator import rk4_step, propagate_batch_numpy
from ..constants import CRITICAL_DISTANCE, WARNING_DISTANCE, ADVISORY_DISTANCE, RE

__all__ = ["Severity", "PcResult", "ConjunctionWarning", "ConjunctionDetector"]


class Severity(StrEnum):
    NONE = "NONE"
    ADVISORY = "ADVISORY"
    WARNING = "WARNING"
    CRITICAL = "CRITICAL"


@dataclass
class PcResult:
    pc: float = 0.0
    sigma_pos_km: float = 0.0
    computed: bool = False


@dataclass
class ConjunctionWarning:
    sat_id: int = 0
    debris_id: int = 0
    current_distance: float = 0.0
    time_to_closest_approach: float = 0.0
    severity: Severity = Severity.NONE
    relative_velocity: List[float] = field(default_factory=lambda: [0.0, 0.0, 0.0])
    pc_result: PcResult = field(default_factory=PcResult)

    @property
    def pc(self) -> float:
        return self.pc_result.pc


def _brent_minimise(f, a: float, b: float, tol: float = 0.1) -> float:
    GOLDEN = 0.3819660
    x = w = v = a + GOLDEN * (b - a)
    fx = fw = fv = f(x)
    d = e = 0.0

    for _ in range(50):
        m = 0.5 * (a + b)
        tol1 = tol * abs(x) + 1e-10
        tol2 = 2.0 * tol1
        if abs(x - m) <= tol2 - 0.5 * (b - a):
            break
        do_golden = True
        if abs(e) > tol1:
            r = (x - w) * (fx - fv)
            q = (x - v) * (fx - fw)
            p = (x - v) * q - (x - w) * r
            q = 2.0 * (q - r)
            if q > 0:
                p = -p
            else:
                q = -q
            r, e = e, d
            if abs(p) < abs(0.5 * q * r) and p > q * (a - x) and p < q * (b - x):
                d = p / q
                u = x + d
                if (u - a) < tol2 or (b - u) < tol2:
                    d = tol1 if x < m else -tol1
                do_golden = False
        if do_golden:
            e = (b - x) if x < m else (a - x)
            d = GOLDEN * e
        u = x + (d if abs(d) >= tol1 else (tol1 if d > 0 else -tol1))
        fu = f(u)
        if fu <= fx:
            if u < x:
                b = x
            else:
                a = x
            v, fv, w, fw, x, fx = w, fw, x, fx, u, fu
        else:
            if u < x:
                a = u
            else:
                b = u
            if fu <= fw or w == x:
                v, fv, w, fw = w, fw, u, fu
            elif fu <= fv or v == x or v == w:
                v, fv = u, fu
    return x


def _chan_pc(
    miss_dist_km: float, sigma_r_km: float, rel_speed_km_s: float, hbr_km: float = 0.01
) -> PcResult:
    """
    Compute collision probability using Chan's method (1997).

    Uses a simplified 2D Gaussian approximation over the encounter plane.
    Assumes spherical hard-body radius and isotropic position uncertainty.

    Args:
        miss_dist_km: Closest approach distance in km.
        sigma_r_km: Combined 1-sigma position uncertainty in km.
        rel_speed_km_s: Relative speed at TCA in km/s.
        hbr_km: Hard-body radius in km (default 10 m).

    Returns:
        PcResult with probability and sigma used.
    """
    r = PcResult()
    r.sigma_pos_km = sigma_r_km
    if sigma_r_km <= 0 or rel_speed_km_s <= 0:
        return r
    x = miss_dist_km / sigma_r_km
    r.pc = min((hbr_km**2 / (2.0 * sigma_r_km**2)) * math.exp(-0.5 * x * x), 1.0)
    r.computed = True
    return r


class ConjunctionDetector:
    """
    Screens satellite-debris pairs for conjunctions.

    Uses a broad-phase KDTree filter (or O(n²) fallback) to find candidate
    pairs, then propagates full histories for all objects and scans for
    minimum-approach distances.

    Pairs exceeding severity thresholds get TCA refinement via Brent minimisation
    of the inter-object distance function.
    """

    def detect(
        self,
        sat_states: List[List[float]],
        debris_states: List[List[float]],
        lookahead_s: float = 86400.0,
        step_s: float = 60.0,
        tle_age_days: float = 1.0,
        mjd0: float = 0.0,
    ) -> List[ConjunctionWarning]:
        """
        Screen all satellite-debris pairs for conjunctions.

        Algorithm:
        1. Broad-phase filtering via KDTree (or O(n²) fallback without SciPy).
        2. Pre-propagate full trajectory for all objects via propagate_batch_full_history.
        3. Coarse temporal sweep: find minimum-approach distance per candidate pair.
        4. Brent refinement of TCA within [tca-step_s, tca+step_s].
        5. Chan collision probability for pairs exceeding ADVISORY threshold.

        Args:
            sat_states: List of satellite ECI state vectors.
            debris_states: List of debris ECI state vectors.
            lookahead_s: Lookahead window in seconds.
            step_s: Temporal sweep step size in seconds.
            tle_age_days: Age of TLE data (used for position uncertainty).
            mjd0: Modified Julian Date at epoch.

        Returns:
            Sorted list of ConjunctionWarning instances (closest first).
        """
        if not sat_states or not debris_states:
            return []

        sat_pos = [s[:3] for s in sat_states]
        deb_pos = [d[:3] for d in debris_states]
        broad_radius = min(15.0 * lookahead_s, 2 * RE)

        try:
            from scipy.spatial import KDTree

            tree = KDTree(deb_pos)
            candidates = tree.query_ball_point(sat_pos, r=broad_radius)
        except ImportError:
            broad_radius2 = broad_radius * broad_radius
            candidates = []
            for s_pos in sat_pos:
                candidate_list = []
                for deb_idx, d_pos in enumerate(deb_pos):
                    dx = s_pos[0] - d_pos[0]
                    dy = s_pos[1] - d_pos[1]
                    dz = s_pos[2] - d_pos[2]
                    if dx * dx + dy * dy + dz * dz <= broad_radius2:
                        candidate_list.append(deb_idx)
                candidates.append(candidate_list)

        n_steps = int(lookahead_s / step_s)
        remainder_s = lookahead_s - n_steps * step_s
        sample_times = [step * step_s for step in range(n_steps + 1)]

        from .accelerator import propagate_batch_full_history

        all_sats = propagate_batch_full_history(sat_states, step_s, n_steps, mjd0=mjd0)
        all_debs = propagate_batch_full_history(
            debris_states, step_s, n_steps, mjd0=mjd0
        )
        if remainder_s > 1e-9:
            remainder_mjd = mjd0 + (n_steps * step_s) / 86400.0 if mjd0 > 0 else 0.0
            sat_tail = propagate_batch_numpy(
                all_sats[-1].tolist(), remainder_s, 1, mjd0=remainder_mjd
            )
            deb_tail = propagate_batch_numpy(
                all_debs[-1].tolist(), remainder_s, 1, mjd0=remainder_mjd
            )
            all_sats = np.concatenate(
                (all_sats, np.array([sat_tail], dtype=np.float64)), axis=0
            )
            all_debs = np.concatenate(
                (all_debs, np.array([deb_tail], dtype=np.float64)), axis=0
            )
            sample_times.append(lookahead_s)

        sigma_pos = 0.3 * math.sqrt(max(tle_age_days, 0.1))

        warnings = []
        for sat_idx, candidate_list in enumerate(candidates):
            for deb_idx in candidate_list:
                min_dist = float("inf")
                tca_coarse = 0.0
                rel_v_at_tca = [0.0, 0.0, 0.0]

                for step, sample_t in enumerate(sample_times):
                    s = all_sats[step][sat_idx]
                    d = all_debs[step][deb_idx]
                    dx = s[0] - d[0]
                    dy = s[1] - d[1]
                    dz = s[2] - d[2]
                    dist = math.sqrt(dx * dx + dy * dy + dz * dz)
                    if dist < min_dist:
                        min_dist = dist
                        tca_coarse = sample_t
                        rel_v_at_tca = [s[3] - d[3], s[4] - d[4], s[5] - d[5]]

                if min_dist >= ADVISORY_DISTANCE:
                    continue

                t_lo = max(0.0, tca_coarse - step_s)
                t_hi = min(lookahead_s, tca_coarse + step_s)

                def dist_at_t(t: float) -> float:
                    n_nearest = int(t / step_s)
                    t_rem = t - n_nearest * step_s
                    elapsed = n_nearest * step_s
                    curr_s = tuple(all_sats[n_nearest][sat_idx])
                    curr_d = tuple(all_debs[n_nearest][deb_idx])
                    if t_rem > 1e-9:
                        curr_s = rk4_step(
                            curr_s, t_rem, mjd0=mjd0, elapsed_seconds=elapsed
                        )
                        curr_d = rk4_step(
                            curr_d, t_rem, mjd0=mjd0, elapsed_seconds=elapsed
                        )
                    dx = curr_s[0] - curr_d[0]
                    dy = curr_s[1] - curr_d[1]
                    dz = curr_s[2] - curr_d[2]
                    return math.sqrt(dx * dx + dy * dy + dz * dz)

                tca_refined = _brent_minimise(dist_at_t, t_lo, t_hi, tol=0.1)
                dist_refined = dist_at_t(tca_refined)

                final_dist = min(min_dist, dist_refined)
                final_tca = tca_refined if dist_refined < min_dist else tca_coarse

                if final_dist < CRITICAL_DISTANCE:
                    severity = Severity.CRITICAL
                elif final_dist < WARNING_DISTANCE:
                    severity = Severity.WARNING
                elif final_dist < ADVISORY_DISTANCE:
                    severity = Severity.ADVISORY
                else:
                    continue

                # Recompute relative velocity at refined TCA
                n_final = int(final_tca / step_s)
                t_final_rem = final_tca - n_final * step_s
                final_s = tuple(all_sats[n_final][sat_idx])
                final_d = tuple(all_debs[n_final][deb_idx])
                if t_final_rem > 1e-9:
                    final_s = rk4_step(final_s, t_final_rem, mjd0=mjd0, elapsed_seconds=n_final * step_s)
                    final_d = rk4_step(final_d, t_final_rem, mjd0=mjd0, elapsed_seconds=n_final * step_s)
                rel_v_at_tca = [final_s[3] - final_d[3], final_s[4] - final_d[4], final_s[5] - final_d[5]]
                rel_speed = math.sqrt(sum(v * v for v in rel_v_at_tca))
                pc = _chan_pc(final_dist, sigma_pos, rel_speed)

                warnings.append(
                    ConjunctionWarning(
                        sat_id=sat_idx,
                        debris_id=deb_idx,
                        current_distance=final_dist,
                        time_to_closest_approach=final_tca,
                        severity=severity,
                        relative_velocity=rel_v_at_tca,
                        pc_result=pc,
                    )
                )

        return warnings

File: cpp/conjunction.cpp
/*
 * cpp/conjunction.cpp — C++ Conjunction Detector
 * ===============================================
 * All-pairs screening with:
 *   - Pre-propagation of all objects (batch, full history) — avoids per-pair
 *     re-propagation during the coarse sweep (was O(ns*nd*steps) RK4 calls,
 *     now O((ns+nd)*steps) + O(ns*nd*steps) distance-only checks).
 *   - Broad-phase spatial culling based on initial pair separation.
 *   - Brent's method for sub-step TCA refinement.
 *   - ADVISORY / WARNING / CRITICAL severity tiers.
 *   - Chan's method Probability of Collision (Pc).
 */
#include "conjunction.h"
#include "propagator.h"
#include <cmath>
#include <algorithm>
#include <cstring>


ConjunctionDetector::ConjunctionDetector() {}

// ── Brent's Method for 1D minimisation ───────────────────────────────────────
// Finds the x in [a,b] minimising f(x) to tolerance tol using Brent's method.
// See: Brent (1973) "Algorithms for Minimization without Derivatives"
template<typename F>
static double brent_minimise(F&& f, double a, double b, double tol = 0.01) {
    constexpr double GOLDEN = 0.3819660;
    double x = a + GOLDEN * (b - a);
    double w = x, v = x;
    double fx = f(x), fw = fx, fv = fx;
    double d = 0.0, e = 0.0;

    for (int iter = 0; iter < 50; ++iter) {
        double m = 0.5 * (a + b);
        double tol1 = tol * std::abs(x) + 1e-10;
        double tol2 = 2.0 * tol1;
        if (std::abs(x - m) <= tol2 - 0.5 * (b - a)) break;

        bool do_golden = true;
        if (std::abs(e) > tol1) {
            double r = (x - w) * (fx - fv);
            double q = (x - v) * (fx - fw);
            double p = (x - v) * q - (x - w) * r;
            q = 2.0 * (q - r);
            if (q > 0) p = -p; else q = -q;
            r = e; e = d;
            if (std::abs(p) < std::abs(0.5 * q * r) &&
                p > q * (a - x) && p < q * (b - x)) {
                d = p / q;
                double u = x + d;
                if ((u - a) < tol2 || (b - u) < tol2)
                    d = (x < m) ? tol1 : -tol1;
                do_golden = false;
            }
        }
        if (do_golden) {
            e = (x < m) ? b - x : a - x;
            d = GOLDEN * e;
        }
        double u = x + ((std::abs(d) >= tol1) ? d : (d > 0 ? tol1 : -tol1));
        double fu = f(u);
        if (fu <= fx) {
            if (u < x) b = x; else a = x;
            v = w; fv = fw; w = x; fw = fx; x = u; fx = fu;
        } else {
            if (u < x) a = u; else b = u;
            if (fu <= fw || w == x) { v = w; fv = fw; w = u; fw = fu; }
            else if (fu <= fv || v == x || v == w) { v = u; fv = fu; }
        }
    }
    return x;
}

// ── Chan's Pc: 2D Gaussian integral (Foster 1992 / Chan 1997 formulation) ────
// Uses the combined covariance at TCA modelled as a diagonal 2D ellipse in the
// miss-distance plane. sigma_r is the combined 1-sigma position uncertainty [km].
// Returns Pc using the series expansion for the circular encounter approximation.
static PcResult chan_pc(double miss_dist_km, double sigma_r_km,
                        double rel_speed_km_s, double hard_body_radius_km = 0.01) {
    PcResult r;
    r.sigma_pos_km = sigma_r_km;
    r.computed = false;
    if (sigma_r_km <= 0 || rel_speed_km_s <= 0) return r;

    double x = miss_dist_km / sigma_r_km;
    double sigma2 = sigma_r_km * sigma_r_km;
    double hbr2 = hard_body_radius_km * hard_body_radius_km;
    double pc = (hbr2 / (2.0 * sigma2)) * std::exp(-0.5 * x * x);
    r.pc = std::min(pc, 1.0);
    r.computed = true;
    return r;
}

std::vector<ConjunctionWarning> ConjunctionDetector::detect(
    const std::vector<StateVector>& sat_states,
    const std::vector<StateVector>& debris_states,
    double lookahead_s,
    double step_s,
    double tle_age_days) const {

    std::vector<ConjunctionWarning> warnings;
    Propagator prop;

    int ns = (int)sat_states.size();
    int nd = (int)debris_states.size();
    if (ns == 0 || nd == 0) return warnings;

    // ── 1. Pre-propagate all objects (batch full history) ──────────────────
    // Flat layout: (steps+1, n, 6) = n_frames × n_objects × 6 doubles.
    // Offset for step t, object i, component k: t * (n * 6) + i * 6 + k.
    int64_t steps = (int64_t)(lookahead_s / step_s);
    int64_t n_frames = steps + 1;

    std::vector<double> init_sats(ns * 6);
    std::vector<double> init_debs(nd * 6);
    for (int i = 0; i < ns; ++i)
        std::memcpy(&init_sats[(size_t)i * 6], sat_states[i].raw(), 6 * sizeof(double));
    for (int i = 0; i < nd; ++i)
        std::memcpy(&init_debs[(size_t)i * 6], debris_states[i].raw(), 6 * sizeof(double));

    std::vector<double> sat_history(n_frames * ns * 6);
    std::vector<double> deb_history(n_frames * nd * 6);

    prop.batch_propagate_full_history(init_sats.data(), ns, step_s, steps,
                                       0.0, 1.0, 2.2, 1.5, false, 0.0,
                                       sat_history.data());
    prop.batch_propagate_full_history(init_debs.data(), nd, step_s, steps,
                                       0.0, 1.0, 2.2, 1.5, false, 0.0,
                                       deb_history.data());

    // ── 2. Broad-phase filter ──────────────────────────────────────────────
    // Cull pairs whose initial separation exceeds the maximum possible
    // relative displacement over the lookahead window.
    double broad_radius = std::min(15.0 * lookahead_s, 2.0 * RE);

    // Position uncertainty grows ~ sqrt(TLE age). Empirical 1-sigma at 1 day: 0.3 km
    double sigma_pos = 0.3 * std::sqrt(std::max(tle_age_days, 0.1));

    // Helper: reconstruct state at arbitrary time t from pre-propagated history.
    // Uses the nearest pre-propagated frame + one RK4 step for the remainder.
    auto state_at_t = [&](const std::vector<double>& history, int n_objects, int obj_idx, double t) -> StateVector {
        int base_step = std::min(static_cast<int64_t>(t / step_s), steps);
        double rem = t - base_step * step_s;
        StateVector s;
        std::memcpy(s.raw(), &history[base_step * (n_objects * 6) + obj_idx * 6], 6 * sizeof(double));
        if (rem > 1e-9) {
            s = prop.propagate(s, rem, 0.0);
        }
        return s;
    };

    // ── 3. Pairwise sweep + Brent refinement ───────────────────────────────
    for (int i = 0; i < ns; ++i) {
        // Broad-phase: sat i initial position
        size_t si = (size_t)i * 6;
        double sx0 = init_sats[si];
        double sy0 = init_sats[si + 1];
        double sz0 = init_sats[si + 2];

        for (int j = 0; j < nd; ++j) {
            size_t dj = (size_t)j * 6;
            double dx0 = sx0 - init_debs[dj];
            double dy0 = sy0 - init_debs[dj + 1];
            double dz0 = sz0 - init_debs[dj + 2];
            if (dx0*dx0 + dy0*dy0 + dz0*dz0 > broad_radius * broad_radius)
                continue;

            // ── Coarse sweep over pre-propagated history ───────────────────
            double min_distance = std::numeric_limits<double>::max();
            double tca_coarse   = 0.0;
            int tca_step        = 0;

            size_t ns6 = (size_t)ns * 6;
            size_t nd6 = (size_t)nd * 6;
            for (int step = 0; step < n_frames; ++step) {
                double* s = &sat_history[(size_t)step * ns6 + si];
                double* d = &deb_history[(size_t)step * nd6 + dj];
                double dx = s[0] - d[0], dy = s[1] - d[1], dz = s[2] - d[2];
                double dist = std::sqrt(dx*dx + dy*dy + dz*dz);

                if (dist < min_distance) {
                    min_distance = dist;
                    tca_coarse   = step * step_s;
                    tca_step     = step;
                }
            }

            if (min_distance >= ADVISORY_DISTANCE) continue;

            // ── Brent refinement ───────────────────────────────────────────
            double t_lo = std::max(0.0, tca_coarse - step_s);
            double t_hi = std::min(lookahead_s, tca_coarse + step_s);

            auto distance_at_t = [&](double t) -> double {
                auto s = state_at_t(sat_history, ns, i, t);
                auto d = state_at_t(deb_history, nd, j, t);
                double dx = s[0]-d[0], dy = s[1]-d[1], dz = s[2]-d[2];
                return std::sqrt(dx*dx + dy*dy + dz*dz);
            };

            double tca_refined = brent_minimise(distance_at_t, t_lo, t_hi, 0.1);
            auto s_tca = state_at_t(sat_history, ns, i, tca_refined);
            auto d_tca = state_at_t(deb_history, nd, j, tca_refined);

            double dx_f = s_tca[0]-d_tca[0];
            double dy_f = s_tca[1]-d_tca[1];
            double dz_f = s_tca[2]-d_tca[2];
            double min_dist_refined = std::sqrt(dx_f*dx_f + dy_f*dy_f + dz_f*dz_f);

            double final_dist = std::min(min_distance, min_dist_refined);
            double final_tca  = (min_dist_refined < min_distance) ? tca_refined : tca_coarse;

            Severity severity = Severity::NONE;
            if      (final_dist < CRITICAL_DISTANCE)  severity = Severity::CRITICAL;
            else if (final_dist < WARNING_DISTANCE)   severity = Severity::WARNING;
            else if (final_dist < ADVISORY_DISTANCE)  severity = Severity::ADVISORY;
            else continue;

            std::array<double, 3> rel_v = {
                s_tca[3] - d_tca[3],
                s_tca[4] - d_tca[4],
                s_tca[5] - d_tca[5]
            };
            double rel_speed = std::sqrt(rel_v[0]*rel_v[0] + rel_v[1]*rel_v[1] + rel_v[2]*rel_v[2]);

            ConjunctionWarning w;
            w.sat_id                   = i;
            w.debris_id                = j;
            w.current_distance         = final_dist;
            w.time_to_closest_approach = final_tca;
            w.severity                 = severity;
            w.relative_velocity        = rel_v;
            w.pc_result                = chan_pc(final_dist, sigma_pos, rel_speed);

            warnings.push_back(w);
        }
    }

    return warnings;
}
